scipy/stats/_distn_infrastructure.py

```diff
@@ -3836,3 +3884,423 @@ def get_distribution_names(namespace_pairs, rv_base_class):
         if isinstance(value, rv_base_class):
             distn_names.append(name)
     return distn_names, distn_gen_names
+
+
+def _combine_bounds(name, bounds_type, user_bounds, shape_domain,
+                    integral):
+    """Intersection of user-defined bounds and distribution PDF/PMF domain"""
+
+    user_bounds = np.atleast_1d(user_bounds)
+
+    if user_bounds.shape != (2,):
+        # `shape_bounds` is already checked, so this can only happen with
+        # `loc_bounds` and `scale_bounds`
+        message = (f"`{bounds_type}` must be a tuple containing exactly "
+                   "two elements: the lower and upper bound of the "
+                   f"`{name}` parameter.")
+        raise ValueError(message)
+
+    if user_bounds[0] > user_bounds[1]:
+        message = (f"There are no values for `{name}` on the interval "
+                   f"{list(user_bounds)}.")
+        raise ValueError(message)
+
+    bounds = (max(user_bounds[0], shape_domain[0]),
+              min(user_bounds[1], shape_domain[1]))
+
+    if integral and (np.ceil(bounds[0]) > np.floor(bounds[1])):
+        message = (f"There are no integer values for `{name}` on the interval "
+                   f"defined by the user-provided bounds and the domain "
+                   "of the distribution.")
+        raise ValueError(message)
+    elif not integral and (bounds[0] > bounds[1]):
+        message = (f"There are no values for `{name}` on the interval "
+                   f"defined by the user-provided bounds and the domain "
+                   "of the distribution.")
+        raise ValueError(message)
+
+    if not np.all(np.isfinite(bounds)):
+        message = (f"The intersection of user-provided bounds for `{name}` "
+                   f"and the domain of the distribution is not finite. Please "
+                   f"provide finite bounds for shape `{name}` in "
+                   f"`{bounds_type}`.")
+        raise ValueError(message)
+
+    return bounds
+
+
+class FitResult:
+    r"""Result of fitting a discrete or continuous distribution to data
+
+    Attributes
+    ----------
+    dist : scipy.stats.rv_continuous or scipy.stats.rv_discrete
+        The distribution of which parameters were fit.
+    data : 1D array_like
+        The data to which the distribution was fit.
+    nllf : callable
+        The negative log-likehood function for the given `data`. Accepts a
+        single tuple containing the shapes, location, and scale of the
+        distribution.
+    params : namedtuple
+        A namedtuple containing the maximum likelihood estimates of the
+        shape parameters, location, and (if applicable) scale of the
+        distribution.
+    objective_val : float
```

Comment:
I'm open to the suggestion of other names. 
I thought `fun` would be ambiguous, considering we're also including `nllf`.

Reply:
Fine with me (I am used to this term) if we add a sentence in the doc explaining the terminology. It optimizes an objective function...

Reply:
We could make this `log_likelihood`, `negative_log_likelihood`, or some acronym (`nll`). Leaving off "function" or "f" distinguishes the value of the function at the optimal solution from the callable function itself.

I expect this to be generalized to other objective functions (i.e. different fitting methods, e.g. method of moments). In this case, would we always want the value of the objective function to be in an attribute with the same name?

Another thought is for there to be methods for evaluating all the possible objective functions, and calling the method with no arguments would return the value of the function at the optimal solution. E.g. the user could call
```python3
res.nllf()
```
to get what is now `res.objective_val`. They can also use `res.nnlf` with arguments to evaluate it with different shape parameters or data.
I think that would be the most elegant way to go, as it provides the most functionality per attribute.

Reply:
Implemented that idea.